text/0000-debug-map-key-value.md

```diff
@@ -0,0 +1,284 @@
+[reference-level-explanation]: #reference-level-explanation
+
+The `key` and `value` methods can be implemented on `DebugMap` by tracking the state of the current entry in a `bool`, and splitting the existing `entry` method into two:
+
+```rust
+pub struct DebugMap<'a, 'b: 'a> {
+    has_key: bool,
+    ..
+}
+
+pub fn debug_map_new<'a, 'b>(fmt: &'a mut fmt::Formatter<'b>) -> DebugMap<'a, 'b> {
+    DebugMap {
+        has_key: false,
+        ..
+    }
+}
+
+impl<'a, 'b: 'a> DebugMap<'a, 'b> {
+    pub fn entry(&mut self, key: &dyn fmt::Debug, value: &dyn fmt::Debug) -> &mut DebugMap<'a, 'b> {
+        self.key(key).value(value)
+    }
+
+    pub fn key(&mut self, key: &dyn fmt::Debug) -> &mut DebugMap<'a, 'b> {
+        self.result = self.result.and_then(|_| {
+            // Make sure there isn't a partial entry
+            if self.has_key {
+                return Err(fmt::Error);
+            }
+
+            if self.is_pretty() {
+                if !self.has_fields {
+                    self.fmt.write_str("\n")?;
+                }
+                let mut slot = None;
+                let mut writer = PadAdapter::wrap(&mut self.fmt, &mut slot);
+                key.fmt(&mut writer)?;
+            } else {
+                if self.has_fields {
+                    self.fmt.write_str(", ")?
+                }
+                key.fmt(self.fmt)?;
+            }
+
+            // Mark that we're in an entry
+            self.has_key = true;
+            Ok(())
+        });
+
+        self
+    }
+
+    pub fn value(&mut self, value: &dyn fmt::Debug) -> &mut DebugMap<'a, 'b> {
+        self.result = self.result.and_then(|_| {
+            // Make sure there is a partial entry to finish
+            if !self.has_key {
+                return Err(fmt::Error);
+            }
+
+            if self.is_pretty() {
+                let mut slot = None;
+                let mut writer = PadAdapter::wrap(&mut self.fmt, &mut slot);
+                writer.write_str(": ")?;
+                value.fmt(&mut writer)?;
+                writer.write_str(",\n")?;
+            } else {
+                self.fmt.write_str(": ")?;
+                value.fmt(self.fmt)?;
+            }
+
+            // Mark that we're not in an entry
+            self.has_key = false;
+            Ok(())
+        });
+
+        self.has_fields = true;
+        self
+    }
+
+    pub fn finish(&mut self) -> fmt::Result {
+        // Make sure there isn't a partial entry
+        if self.has_key {
+            return Err(fmt::Error);
+        }
+
+        self.result.and_then(|_| self.fmt.write_str("}"))
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The proposed `key` and `value` methods are't immediately useful for `Debug` implementors that are able to call `entry` instead. This creates a decision point where there wasn't one before. The proposed implementation is also going to be less efficient than the one that exists now because it introduces a few conditionals.
```

Comment:
Regarding efficiency, I would suggest testing this out with a PR and then doing a timer build to see how much perf regresses (that just affects compile times but it's better than nothing). With that data we can evaluate the drawback better.

Reply:
I opened up a PR (https://github.com/rust-lang/rust/pull/60458) that we can use to check correctness and get some timings :+1:

I don't expect the difference to be significant, since it amounts to a few conditionals but it'll be good to verify!

Reply:
Alrighty, I posted a quick benchmark in the PR, but also inlining the results here:

| Before | After |
| ------ | ----- |
| 36 ns/iter (+/- 7) | 44 ns/iter (+/- 2) |